worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java

```diff
@@ -175,21 +339,43 @@ private void processReject(long id, boolean requeue)
         }
     }
 
-    /**
-     * Republish the delivery to the retry queue with the retry count stamped in the headers.
-     *
-     * @param delivery the redelivered message
-     */
-    private void republishClassicRedelivery(final Delivery delivery, final int retries) {
-
-        final RabbitTaskInformation taskInformation = 
-                new RabbitTaskInformation(String.valueOf(delivery.getEnvelope().getDeliveryTag()));
+    private void republishClassicRedelivery(
+        final String deliveryQueue,
+        final long inboundMessageId,
+        final byte[] serializedTaskMessage,
+        final Optional<String> taskMessageStorageRefOpt,
+        final int retries,
+        final TrackingInfo tracking
+    ) throws InvalidDeliveryException
+    {
+        final String trackingJobTaskId = tracking != null ? tracking.getJobTaskId() : "untracked";
+        final RabbitTaskInformation taskInformation = new RabbitTaskInformation(
+            String.valueOf(inboundMessageId), false, Optional.of(trackingJobTaskId));
         LOG.debug("Received redelivered message with id {}, retry count {}, retry limit {}, republishing to retry queue",
-                delivery.getEnvelope().getDeliveryTag(), retryLimit, retries + 1);
-        final Map<String, Object> headers = new HashMap<>();
-        headers.put(RabbitHeaders.RABBIT_HEADER_CAF_WORKER_RETRY, String.valueOf(retries + 1));
+                  inboundMessageId, retryLimit, retries + 1);
+        final Map<String, Object> publishHeaders = new HashMap<>();
+        publishHeaders.put(RabbitHeaders.RABBIT_HEADER_CAF_WORKER_RETRY, String.valueOf(retries + 1));
+        taskMessageStorageRefOpt.ifPresent(s -> publishHeaders.put(RABBIT_HEADER_CAF_PAYLOAD_OFFLOADING_STORAGE_REF, s));
         taskInformation.incrementResponseCount(true);
-        publisherEventQueue.add(new WorkerPublishQueueEvent(delivery.getMessageData(), retryRoutingKey, 
-                taskInformation, headers));
+        if(taskMessageStorageRefOpt.isPresent()) {
+            if (!retryRoutingKey.equals(deliveryQueue)) {
+                try {
+                    dataStore.store(dataStore.retrieve(taskMessageStorageRefOpt.get()), taskMessageStorageRefOpt.get().replace(deliveryQueue, retryRoutingKey));
+                } 
+                catch (final ReferenceNotFoundException e) {
+                    throw new InvalidDeliveryException("Original reference not found when relocating TaskData", inboundMessageId, e);
+                }
+                catch (final DataStoreException e) {
+                    LOG.error("Failed to relocate offloaded payload for message id {} from {} to {}",
+                            inboundMessageId, deliveryQueue, retryRoutingKey, e);
+                    disconnectCallback.run();
+                }
```

Comment:
Does retrieve only throw the ReferenceNotFoundException?
Don't we need to update the publish headers here too?